### PR TITLE
[dynamo 3.11] fix segfault when printing stack trace

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99180
* __->__ #99934

Dynamo will frequently segfault when attempting to print stack traces. We fix this by:
- Fixing stack size calculations, as we did not account for exception tables
- Creating shadow execution frames in a way that more closely resembles what CPython does to create its execution frames

Dynamo/inductor-wrapped pytorch tests are enabled up the stack - those need to be green before this PR can be merged.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @soumith @anijain2305 @Xia-Weiwen @desertfire

### CODE REVIEW

File: torch/csrc/dynamo/eval_frame.c [hunk 1]
```diff
@@ -542,10 +549,20 @@ inline static PyObject* eval_custom_code(
     Py_XINCREF(fastlocals_old[nlocals_old + i]);
     fastlocals_new[nlocals_new + i] = fastlocals_old[nlocals_old + i];
   }
+
+  #endif
+
+  result = eval_frame_default(tstate, shadow, throw_flag);
+
+  #if IS_PYTHON_3_11_PLUS
+ cleanup:
```
Comment (albanD) [inline torch/csrc/dynamo/eval_frame.c:558]: I know CPython uses a lot of gotos but we usually don't. There are 0 usage of goto across torch/ aten/ c10/ today.
I think it is ok to duplicate these 3 cleanup lines above in this case.

File: torch/csrc/dynamo/eval_frame.c [hunk 2]
```diff
@@ -461,57 +461,55 @@ inline static PyObject* eval_custom_code(
...
-  #else
-  THP_EVAL_API_FRAME_OBJECT* shadow = shadow_obj;
-  #endif
+  PyFunctionObject *func = _PyFunction_CopyWithNewCode((PyFunctionObject *) frame->f_func, code);
+  if (func == NULL) {
+    return NULL;
+  }
+
+  size_t size = code->co_nlocalsplus + code->co_stacksize + FRAME_SPECIALS_SIZE;
+  THP_EVAL_API_FRAME_OBJECT* shadow = malloc(size);
```
Comment (albanD) [inline torch/csrc/dynamo/eval_frame.c:479]: I think we want a more detailed comment on why we do that and why it is ok here

File: torch/csrc/dynamo/eval_frame.c [hunk 3]
```diff
@@ -461,57 +461,55 @@ inline static PyObject* eval_custom_code(
...
+  Py_INCREF(func);
+  // consumes reference to func
+  _PyFrame_InitializeSpecials(shadow, func, NULL, code->co_nlocalsplus);
+
   PyObject** fastlocals_old = frame->localsplus;
   PyObject** fastlocals_new = shadow->localsplus;
 
-  // copy from old fastlocals to new fastlocals:
+  for (int i = 0; i < code->co_nlocalsplus; i++) {
+    fastlocals_new[i] = NULL;
```
Comment (albanD) [inline torch/csrc/dynamo/eval_frame.c:493]: Why is this needed? Is that in case you don't re-populate the whole array you don't want it to be uninitialized?

**Review by albanD — COMMENTED** [review]
Sounds pretty good.
Are there specific tests we should re-enable following this? Or add new one-off tests to show what this fixed?

File: torch/csrc/dynamo/eval_frame.c (hunk 3 above)
Comment (williamwen42) [reply torch/csrc/dynamo/eval_frame.c:493]: Yes. Every element in the localsplus array will be Py_XINCREF'd.

**Comment by williamwen42** [comment]
I'll add a test that launches a subprocess running code with an unhandled exception. Other tests will be enabled in the next PR.

File: test/dynamo/test_misc.py [hunk 4]
```diff
@@ -4697,6 +4698,27 @@ def test_remove_dead_code_with_exn_table_entries(self):
...
+
+opt_fn = torch.compile(backend="eager")(f)
+opt_fn(torch.ones(2))
+"""
+        proc = subprocess.run(
+            ["python3", "-c", code],
+            capture_output=True,
+        )
+        # process did not segfault, C assert fail (abort), etc.
+        self.assertGreater(proc.returncode, 0)
```
Comment (albanD) [inline test/dynamo/test_misc.py:4719]: Segfault would actually give you a non-zero code?

File: torch/csrc/dynamo/eval_frame.c (hunk 3 above)
Comment (albanD) [reply torch/csrc/dynamo/eval_frame.c:495]: Can you have a comment here explaining that please?

**Review by albanD — APPROVED** [review]
nit for comment bug sounds good otherwise!

File: test/dynamo/test_misc.py [hunk 5]
```diff
@@ -4698,6 +4699,40 @@ def test_remove_dead_code_with_exn_table_entries(self):
...
+def f(a):
+    a += 1
+    raise RuntimeError("smoge")
+    return a
+
+opt_fn = torch._dynamo.optimize("eager")(f)
+opt_fn(torch.ones(2))
+"""
+        proc = subprocess.run(
+            ["python3", "-c", code],
```
Comment (malfet) [inline test/dynamo/test_misc.py:4717]: Please always use `sys.executable` as `python3` can not be the version of Python this test was executed with. Also, why not use `multiprocessing`
```suggestion
            [sys.executable, "-c", code],
```

File: test/dynamo/test_misc.py (hunk 4 above)
Comment (williamwen42) [reply test/dynamo/test_misc.py:4721]: Exits due to signals (segfault, abort, etc.) will have negative return code.

Comment (albanD) [reply test/dynamo/test_misc.py:4721]: I don't think that this is generally true?
This is not true on my Fedora install at least:
```
$ python foo.py 
Segmentation fault (core dumped)
$ echo $?
139
```

Comment (williamwen42) [reply test/dynamo/test_misc.py:4721]: This behavior is specific to python: https://docs.python.org/3/library/subprocess.html#subprocess.CompletedProcess.returncode

Comment (williamwen42) [reply test/dynamo/test_misc.py:4783]: We'll skip the test if not on POSIX

File: test/dynamo/test_misc.py [hunk 6]
```diff
@@ -4721,6 +4722,77 @@ def test_remove_dead_code_with_exn_table_entries(self):
         self.assertEquals(tab[0].end, 4)
         self.assertEquals(tab[0].target, 6)
 
+    @unittest.skipIf(os.name != "posix", "return code check is POSIX only")
```
Comment (malfet) [inline test/dynamo/test_misc.py:4725]:
```suggestion
   @cpuOnly
    @unittest.skipIf(os.name != "posix", "return code check is POSIX only")
```

**Review by malfet — APPROVED** [review]
Can you please gate those checks to CPU only shards, as forking on GPUs runner is prohibitively expensive, but the crash is reproducible on CPU machines, isn't it?)

**Review by jansel — APPROVED** [review]

File: test/dynamo/test_misc.py (hunk 6 above)
Comment (williamwen42) [reply test/dynamo/test_misc.py]: Turns out it's enough to use `traceback.format_exc()` to trigger the segfault - no need to spawn a separate process.

File: test/dynamo/test_misc.py (hunk 4 above)
Comment (williamwen42) [reply test/dynamo/test_misc.py]: Update: test changed to trigger segfault without spawning a subprocess. POSIX check no longer required.

**Comment by williamwen42** [comment]
Dynamo-wrapped pytorch tests look good for the most part at the moment up the stack (https://github.com/pytorch/pytorch/pull/99180). Will now merge this PR.
